prql/src/semantic/reporting.rs

```diff
@@ -0,0 +1,84 @@
+use std::ops::Range;
+
+use anyhow::{Ok, Result};
+use ariadne::{Color, Label, Report, ReportBuilder, ReportKind, Source};
+
+use super::{Declaration, SemanticAnalyzer, VarDec};
+use crate::ast::FuncDef;
+use crate::internals::{AstFold, Node};
+
+pub fn print(analyzer: &SemanticAnalyzer, source_id: String, source: String) {
+    let mut report = Report::build(ReportKind::Custom("Info", Color::Blue), &source_id, 0);
+
+    let source = Source::from(source);
+
+    // label all idents and function calls
+    let mut labeller = Labeller {
+        analyzer,
+        source: &source,
+        source_id: &source_id,
+        report: &mut report,
+    };
+    // traverse ast
+    labeller.fold_node(analyzer.get_ast().clone()).unwrap();
+    // traverse declarations
+    for (d, _) in &analyzer.declarations {
+        match d {
+            Declaration::Variable(VarDec { declaration: n, .. })
+            | Declaration::Function(FuncDef { body: n, .. }) => {
+                labeller.fold_node(*(*n).clone()).unwrap();
+            }
+            Declaration::Table(_) => todo!(),
+        }
+    }
+
+    // label all declarations
+    // for (dec, span) in &analyzer.declarations {
+    //     if let Some(span) = span {
+    //         report.add_label(
+    //             Label::new((source_id.clone(), Range::from(*span)))
+    //                 .with_message(dec.to_string()),
+    //         );
+    //     }
+    // }
+
+    report.finish().print((source_id, source)).unwrap();
+}
+
+/// Traverses AST and add labels for each of the idents and function calls
+struct Labeller<'a> {
```

Comment:
Despite being from London, I think we should probably keep the American spelling :)

```suggestion
struct Labeler<'a> {
```